homeassistant/components/esphomelib/__init__.py

```diff
@@ -0,0 +1,338 @@
+"""Support for esphomelib devices."""
+import asyncio
+import logging
+from typing import Any, Dict, List, Union
+
+import voluptuous as vol
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_HOST, CONF_PASSWORD, CONF_PORT, \
+    EVENT_HOMEASSISTANT_STOP
+from homeassistant.core import callback
+from homeassistant.helpers.dispatcher import async_dispatcher_connect, \
+    async_dispatcher_send
+from homeassistant.helpers.entity import Entity
+from homeassistant.helpers.typing import HomeAssistantType
+
+DOMAIN = 'esphomelib'
+REQUIREMENTS = ['aioesphomeapi==1.0.0']
+
+# The key to store the APIClient in hass.data
+DATA_CLIENT = 'client'
+# the key to store the scheduled reconnect task in hass.data
+DATA_RECONNECT_TASK = 'reconnect'
+# The key to store the available state of the device (True/False)
+DATA_AVAILABLE = 'available'
+# The key to store a DeviceInfo instance of the device
+DATA_DEVICE_INFO = 'device_info'
+# Dispatcher key to distribute state changes to entities
+DISPATCHER_ON_STATE = 'esphomelib_{}_on_state'
+# Dispatcher key to distribute entity listings on ESP boot
+DISPATCHER_ON_LIST = 'esphomelib_{}_on_list'
+# Dispatcher key to distribute changes to the availability of an ESP
+DISPATCHER_ON_DEVICE_UPDATE = 'esphomelib_{}_on_device_update'
+# The HA component types this integration supports
+HA_COMPONENTS = ['sensor']
+
+_LOGGER = logging.getLogger(__name__)
+
+# No config schema - only configuration entry
+CONFIG_SCHEMA = vol.Schema({}, extra=vol.ALLOW_EXTRA)
+
+
+async def async_setup_entry(hass: HomeAssistantType, entry: ConfigEntry):
+    """Set up the esphomelib component."""
+    from aioesphomeapi.client import APIClient, APIConnectionError
+
+    hass.data.setdefault(DOMAIN, {})
+
+    host = entry.data[CONF_HOST]
+    port = entry.data[CONF_PORT]
+    password = entry.data[CONF_PASSWORD]
+
+    cli = APIClient(hass.loop, host, port, password)
+    await cli.start()
+
+    async def on_stop(event):
+        """Cleanup the socket client on HA stop."""
+        await _cleanup_instance(hass, entry)
+
+    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, on_stop)
+
+    # Store client in per-config-entry hass.data
+    entry_data = get_entry_data(hass, entry)
+    entry_data[DATA_CLIENT] = cli
+    entry_data[DATA_AVAILABLE] = False
+
+    try_connect = await _setup_auto_reconnect_logic(
+        hass, cli, entry.entry_id, host)
```

Comment:
If we pass the entry here instead, we could make the `get_entry_data` more strict about what argument types it should accept.